protocol/0087-VAMM-automated_market_maker.md

```diff
@@ -0,0 +1,165 @@
+
+To calculate this, the interface will need the `starting price` $p_s$, `ending price` $p_e$, `upper price of the current range` $p_u$ (`upper price` if `P < 0` else `lower price`) and the `L` value for the current range. At `P = 0` use the values for the range which the volume change will cause the position to move into.
+
+We then need to calculate the implied position at `starting price` and `ending price` and return the difference. 
+
+For a given price $p$ we calculate implied position $P_i$ with 
+
+$$
+P_i = L * \frac{\sqrt{p_u} - \sqrt{p}}{\sqrt{p} \cdotp \sqrt{p_u}}
+$$
+
+Then simply return the absolute difference between these two prices.
+
+## Matching Process (To merge with 0068-MATC once confirmed)
+
+For all incoming active orders, the matching process will coordinate between the on- and off-book sources of liquidity. When an order comes in which may immediately trade (there are not already resting orders of the same type for the best applicable price) the following steps should be followed. If at any point the order's full volume has traded the process is immediately halted:
+
+  1. For the first applicable price level, all on-book orders should be checked. Any volume at this price level which can be met through on-book orders will then trade. 
+  1. For any `remaining volume`, the AMMs will be checked. This requires an algorithm to ensure we do not have to check every price level individually:
+     1. Call the current price level `current price`
+     1. Check the price level which has the next resting on-book order, set this to be the `outer price` for the check.
+     1. Check all active AMMs, querying their quote price API with the smallest trade unit on the market in the direction of trading (if the incoming order is a `buy`, query the AMM's `ask`, or vice versa). Retain those where this price < `outer price`
+     1. Within these, select either the minimum `upper price` (if the incoming order is a buy) or the maximum `lower price` (if the incoming order is a sell), call this `amm bound price`. This is the range where all of these AMMs are active. Finally, select either the minimum (for a buy) or maximum (for a sell) between `amm bound price` and `outer price`. From this form an interval `current price, outer price`. 
+     1. Now, for each AMM within this range, calculate the volume of trading required to move each from the `current price` to the `outer price`. Call the sum of this volume `total volume`.
+     1. If `remaining volume <= total volume` split trades between the AMMs according to their proportional contribution to `total volume` (e.g. larger liquidity receives a higher proportion of the trade). This ensures their mid prices will move equally (TODO: Is trade splitting more involved than this?).
```

Comment:
we should probably mention the `maker_fee` going to the respective AMMs for the volume they contributed.